theseus/geometry/se3.py

```diff
@@ -250,67 +254,80 @@ def exp_map(tangent_vector: torch.Tensor) -> "SE3":
+            not_near_pi.logical_and(near_zero), 1 + sine**2 / 6, theta / sine
+        ).view(-1, 1)
+        rot_not_near_pi = rotation_base * scale
+
+        # theta is ~pi
+        aux = torch.arange(self.shape[0], device=self.device)
+        ddiag = torch.diagonal(self.data, dim1=1, dim2=2)
         # Find the index of major coloumns and diagonals
         major = torch.logical_and(
             ddiag[:, 1] > ddiag[:, 0], ddiag[:, 1] > ddiag[:, 2]
         ) + 2 * torch.logical_and(ddiag[:, 2] > ddiag[:, 0], ddiag[:, 2] > ddiag[:, 1])
-        ret[near_pi, 3:] = self[near_pi, major, :3]
-        ret[near_pi, major + 3] -= cosine[near_pi]
-        ret[near_pi, 3:] *= (theta[near_pi] ** 2 / (1 - cosine[near_pi])).view(-1, 1)
-        ret[near_pi, 3:] /= ret[near_pi, major + 3].sqrt().view(-1, 1)
+        rot_near_pi = self[aux, major, :3]
+        rot_near_pi[aux, major] -= cosine
+        rot_near_pi *= (theta**2 / (1 - cosine)).view(-1, 1)
+        rot_near_pi /= rot_near_pi[aux, major].sqrt().view(-1, 1)
 
-        # Compute the translation
-        near_zero = theta < NEAR_ZERO_EPS
+        return torch.where(not_near_pi.view(-1, 1), rot_not_near_pi, rot_near_pi)
+
+    def _compute_log_map_translation(
+        self,
+        sine: torch.Tensor,
+        cosine: torch.Tensor,
+        theta: torch.Tensor,
+        rotation: torch.Tensor,
+    ) -> torch.Tensor:
+        theta2 = theta**2
+
+        near_zero = theta < SE3.NEAR_ZERO_EPS
         sine_theta = sine * theta
         two_cosine_minus_two = 2 * cosine - 2
-        two_cosine_minus_two_nz = torch.where(near_zero, non_zero, two_cosine_minus_two)
-        theta2_nz = torch.where(near_zero, non_zero, theta2)
-        a = torch.where(
-            near_zero, 1 - theta2 / 12, -sine_theta / two_cosine_minus_two_nz
-        )
+        a = torch.where(near_zero, 1 - theta2 / 12, -sine_theta / two_cosine_minus_two)
         b = torch.where(
             near_zero,
             1.0 / 12 + theta2 / 720,
-            (sine_theta + two_cosine_minus_two) / (theta2_nz * two_cosine_minus_two_nz),
+            (sine_theta + two_cosine_minus_two) / (theta2 * two_cosine_minus_two),
         )
 
-        translation = self[:, :, 3].view(-1, 3, 1)
-        tangent_vector_ang = ret[:, 3:].view(-1, 3, 1)
-        ret[:, :3] = a.view(-1, 1) * self[:, :, 3]
-        ret[:, :3] -= 0.5 * torch.cross(ret[:, 3:], self[:, :, 3])
-        ret[:, :3] += b.view(-1, 1) * (
-            tangent_vector_ang @ (tangent_vector_ang.transpose(1, 2) @ translation)
+        tangent_vector_ang = rotation.view(-1, 3, 1)
+        translation = a.view(-1, 1) * self[:, :, 3]
+        translation -= 0.5 * torch.cross(rotation, self[:, :, 3])
+        translation += b.view(-1, 1) * (
+            tangent_vector_ang
+            @ (tangent_vector_ang.transpose(1, 2) @ self[:, :, 3].view(-1, 3, 1))
         ).view(-1, 3)
+        return translation
 
-        return ret
+    def _log_map_impl(self) -> torch.Tensor:
```

Comment:
We should make sure whether we need similar fixes in other functions or classes (SE2/SO2/SO3).

Reply:
I checked `SE2` and `SO2` and they should be `OK`.